src/hmacsha256p256/mod.rs

```diff
@@ -0,0 +1,241 @@
+                _ => Err(Error::InternalError),
+            }
+        }
+    }
+
+    impl From<InjectAnyKey> for HmacSha256P256Request {
+        fn from(request: InjectAnyKey) -> Self {
+            Self::InjectAnyKey(request)
+        }
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[allow(missing_docs)]
+pub enum HmacSha256P256Reply {
+    DeriveFromHash(reply::DeriveFromHash),
+    InjectAnyKey(reply::InjectAnyKey),
+}
+
+mod reply {
+    use serde::{Deserialize, Serialize};
+    use trussed::{types::KeyId, Error};
+
+    use super::*;
+
+    #[derive(Debug, Deserialize, Serialize)]
+    #[non_exhaustive]
+    pub struct DeriveFromHash {
+        pub key: Option<KeyId>,
+    }
+
+    impl TryFrom<HmacSha256P256Reply> for DeriveFromHash {
+        type Error = Error;
+        fn try_from(reply: HmacSha256P256Reply) -> Result<Self, Self::Error> {
+            match reply {
+                HmacSha256P256Reply::DeriveFromHash(reply) => Ok(reply),
+                _ => Err(Error::InternalError),
+            }
+        }
+    }
+
+    impl From<DeriveFromHash> for HmacSha256P256Reply {
+        fn from(reply: DeriveFromHash) -> Self {
+            Self::DeriveFromHash(reply)
+        }
+    }
+
+    #[derive(Debug, Deserialize, Serialize)]
+    pub struct InjectAnyKey {
+        pub key: Option<KeyId>,
+    }
+
+    impl TryFrom<HmacSha256P256Reply> for InjectAnyKey {
+        type Error = Error;
+        fn try_from(reply: HmacSha256P256Reply) -> Result<Self, Self::Error> {
+            match reply {
+                HmacSha256P256Reply::InjectAnyKey(reply) => Ok(reply),
+                _ => Err(Error::InternalError),
+            }
+        }
+    }
+
+    impl From<InjectAnyKey> for HmacSha256P256Reply {
+        fn from(reply: InjectAnyKey) -> Self {
+            Self::InjectAnyKey(reply)
+        }
+    }
+}
+
+impl Extension for HmacSha256P256Extension {
+    type Request = HmacSha256P256Request;
+    type Reply = HmacSha256P256Reply;
+}
+
+pub fn derive_key_from_hash(
+    keystore: &mut impl Keystore,
+    request: &request::DeriveFromHash,
+) -> Result<reply::DeriveFromHash, Error> {
+    use hmac::{Hmac, Mac};
+    type HmacSha256P256 = Hmac<sha2::Sha256>;
+
+    let key_id = request.key;
+    let key = keystore.load_key(key::Secrecy::Secret, None, &key_id)?;
+    let shared_secret = key.material;
+
+    let mut mac =
+        HmacSha256P256::new_from_slice(shared_secret.as_ref()).map_err(|_| Error::InternalError)?;
+
+    if let Some(data) = &request.data {
+        mac.update(data);
+    }
+    let derived_key: [u8; 32] = mac
+        .finalize()
+        .into_bytes()
+        .try_into()
+        .map_err(|_| Error::InternalError)?;
+    let key_id = keystore.store_key(
+        request.location,
+        key::Secrecy::Secret,
+        key::Kind::P256, // TODO use mechanism/kind from the request
+        &derived_key,
+    )?;
+    Ok(reply::DeriveFromHash { key: Some(key_id) })
```

Comment:
Are you sure that this necessarily creates a key that is valid?

The key generation is a bit more complex than just using random bytes. This is the procedure we currently use in trussed: https://docs.rs/p256-cortex-m4/0.1.0-alpha.6/src/p256_cortex_m4/cortex_m4.rs.html#81-93.

Especially, the private key has to be within the $[1; n-1]$ where $n$ is the prime order of P-256.

Reply:
That's a correct observation - there are no checks for key validation as of now. I was thinking about adding the same checks like done in the linked implementation, but I am gravitating now more towards using curve25519 instead, and remove the p256 support for derivation, unless proved to be safe to use in this way.
I've just noticed there is no ticket for that - to be done soon.
Connected: https://github.com/trussed-dev/trussed/pull/44#issuecomment-1280559597

Reply:
For `EdDSA` this will work but for X25519 (ECDH) there is the same issue, except the constraints on the scalar are even worse.

Reply:
One solution could be to simply use the result of the hash to seed a `ChaCha8Rng` and use it to generate the key. We would need it to commit to having the key generation algorithm be stable so that the same RNG always gives the same key.